src/Data/Functor/Foldable/TH.hs

```diff
@@ -70,12 +74,59 @@ import Paths_recursion_schemes (version)
 -- as we don't try to do better than
 -- <https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/glasgow_exts.html#deriving-functor-instances GHC's DeriveFunctor>.
 --
-makeBaseFunctor :: Name -> DecsQ
-makeBaseFunctor = makeBaseFunctorWith baseRules
+-- Some doctests:
+--
+-- >>> data Expr a = Lit a | Add (Expr a) (Expr a) | Expr a :* [Expr a]
+-- >>> ; makeBaseFunctor ''Expr
+--
+-- >>> :t AddF
+-- AddF :: r -> r -> ExprF a r
+--
+-- >>> data Rose f a = Rose a (f (Rose f a))
+-- >>> ; makeBaseFunctor [d| instance Functor f => Recursive (Rose f a) |]
+--
+-- >>> :t RoseF
+-- RoseF :: a -> f r -> RoseF f a r
+--
+-- >>> let rose = Rose 1 (Just (Rose 2 (Just (Rose 3 Nothing))))
+-- >>> cata (\(RoseF x f) -> x + maybe 0 id f) rose
+-- 6
+--
+class MakeBaseFunctor a where
+    makeBaseFunctor :: a -> DecsQ
+
+instance MakeBaseFunctor a => MakeBaseFunctor [a] where
+    makeBaseFunctor a = fmap concat (T.mapM makeBaseFunctor a)
+
+instance MakeBaseFunctor a => MakeBaseFunctor (Q a) where
+    makeBaseFunctor a = makeBaseFunctor =<< a
+
+instance MakeBaseFunctor Name where
+    makeBaseFunctor = makeBaseFunctorWith baseRules
+
+instance MakeBaseFunctor Dec where
+#if MIN_VERSION_template_haskell(2,11,0)
+    makeBaseFunctor (InstanceD overlaps ctx classHead []) = do
+        let instanceFor = InstanceD overlaps ctx
+#else
+    makeBaseFunctor (InstanceD ctx classHead []) = do
+        let instanceFor = InstanceD ctx
+#endif
+        case classHead of
+          ConT u `AppT` t | u == recursiveTypeName -> do
+              name <- headOfType t
+              di <- reifyDatatype name
+--              i <- fmap instanceFor $ mapM id
+--                  [ instanceDecFor di
+--                  ]
+              makePrimForDI baseRules (Just $ \n -> instanceFor (ConT n `AppT` t)) di
+          _ -> fail $ "makeBaseFunctor: expected an instance head like `Recursive (C a b ...)`, got " ++ show classHead
```

Comment:
Perhaps say "expected an instance head like `ctxt => Recursive (C a b ...)`" to make it clear that instance contexts are permitted.

Reply:
done